cpp/src/arrow/compute/exec/tpch_node_test.cc

```diff
@@ -0,0 +1,631 @@
+                 const std::unordered_set<util::string_view>& possibilities) {
+  int64_t length = d.length();
+  const char* col = reinterpret_cast<const char*>(d.array()->buffers[1]->data());
+  for (int64_t i = 0; i < length; i++) {
+    const char* row = col + i * byte_width;
+    int32_t row_len = 0;
+    while (row[row_len] && row_len < byte_width) row_len++;
+    util::string_view view(row, row_len);
+    ASSERT_TRUE(possibilities.find(view) != possibilities.end())
+        << view << " is not a valid string.";
+  }
+}
+
+// Counts the number of instances of each integer
+void CountInstances(std::unordered_map<int32_t, int32_t>& counts, const Datum& d) {
+  int64_t length = d.length();
+  const int32_t* nums = reinterpret_cast<const int32_t*>(d.array()->buffers[1]->data());
+  for (int64_t i = 0; i < length; i++) counts[nums[i]]++;
+}
+
+// For the S_COMMENT column, some of the columns must be modified to contain
+// "Customer...Complaints" or "Customer...Recommends". This function counts the number of
+// good and bad comments.
+void CountModifiedComments(const Datum& d, int& good_count, int& bad_count) {
+  int64_t length = d.length();
+  const int32_t* offsets =
+      reinterpret_cast<const int32_t*>(d.array()->buffers[1]->data());
+  const char* str = reinterpret_cast<const char*>(d.array()->buffers[2]->data());
+  // Length of S_COMMENT is at most 100
+  char tmp_string[101];
+  for (int64_t i = 0; i < length; i++) {
+    const char* row = str + offsets[i];
+    int32_t row_length = offsets[i + 1] - offsets[i];
+    std::memset(tmp_string, 0, sizeof(tmp_string));
+    std::memcpy(tmp_string, row, row_length);
+    char* customer = std::strstr(tmp_string, "Customer");
+    char* recommends = std::strstr(tmp_string, "Recommends");
+    char* complaints = std::strstr(tmp_string, "Complaints");
+    if (customer) {
+      ASSERT_TRUE((recommends != nullptr) ^ (complaints != nullptr));
+      if (recommends) good_count++;
+      if (complaints) bad_count++;
+    }
+  }
+}
+
+TEST(TpchNode, ScaleFactor) {
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  std::shared_ptr<ExecPlan> plan = *ExecPlan::Make(&ctx);
+  TpchGen gen = *TpchGen::Make(plan.get(), 0.25f);
+  ExecNode* table = *gen.Supplier();
+  AsyncGenerator<util::optional<ExecBatch>> sink_gen;
+  Declaration sink("sink", {Declaration::Input(table)}, SinkNodeOptions{&sink_gen});
+  std::ignore = *sink.AddToPlan(plan.get());
+  auto fut = StartAndCollect(plan.get(), sink_gen);
+  auto res = *fut.MoveResult();
+
+  int64_t kExpectedRows = 2500;
+  int64_t num_rows = 0;
+  for (auto& batch : res) num_rows += batch.length;
+  ASSERT_EQ(num_rows, kExpectedRows);
+}
+
+TEST(TpchNode, Supplier) {
+  ExecContext ctx(default_memory_pool(), arrow::internal::GetCpuThreadPool());
+  std::shared_ptr<ExecPlan> plan = *ExecPlan::Make(&ctx);
+  TpchGen gen = *TpchGen::Make(plan.get());
+  ExecNode* table = *gen.Supplier();
```

Comment:
Here is as well, check for unexpected errors instead of potentially crashing the test process...
```suggestion
  ASSERT_OK_AND_ASSIGN(std::shared_ptr<ExecPlan> plan, ExecPlan::Make(&ctx));
  ASSERT_OK_AND_ASSIGN(TpchGen gen, TpchGen::Make(plan.get(), 0.25f));
```